lib/backend/k8s/resources/profile.go

```diff
@@ -148,7 +251,212 @@ func (c *profileClient) Watch(ctx context.Context, list model.ListInterface, rev
+				log.Warn("Profile, namespace watch channel closed.")
+				e = api.WatchEvent{
+					Type: api.WatchError,
+					Error: cerrors.ErrorWatchTerminated{
+						ClosedByRemote: true,
+						Err:            errors.New("Profile namespace watch channel closed."),
+					},
+				}
+			}
+			log.Debug("Processing Namespace event")
+			isNsEvent = true
+
+		case e, ok = <-pw.k8sSAWatch.ResultChan():
+			if !ok {
+				// We shouldn't get a closed channel without first getting a terminating error,
+				// so write a warning log and convert to a termination error.
+				log.Warn("Profile, serviceaccount watch channel closed.")
+				e = api.WatchEvent{
+					Type: api.WatchError,
+					Error: cerrors.ErrorWatchTerminated{
+						ClosedByRemote: true,
+						Err:            errors.New("Profile serviceaccount watch channel closed."),
+					},
+				}
+			}
+			log.Debug("Processing SeriveAccount event")
+			isNsEvent = false
+
+		case <-pw.context.Done(): //user cancel
+			log.Info("Process watcher done event in kdd client")
+			return
+		}
+
+		// Update the resource version of the Object in the watcher. The version returned on a watch
+		// event needs to be such that the Watch client can resume watching when a watch fails.
+		// The watch client expects a slash separated list of resource versions in the format
+		// <NS Revision/SA Revision>.
+		var value interface{}
+		switch e.Type {
+		case api.WatchModified, api.WatchAdded:
+			value = e.New.Value
+		case api.WatchDeleted:
+			value = e.Old.Value
+		}
+
+		if value != nil {
+			oma, ok := value.(metav1.ObjectMetaAccessor)
+
+			if !ok {
+				log.WithField("event", e).Error(
+					"Resource returned from watch does not implement ObjectMetaAccessor interface")
+				e = api.WatchEvent{
+					Type: api.WatchError,
+					Error: cerrors.ErrorWatchTerminated{
+						ClosedByRemote: true,
+						Err:            errors.New("Profile value does not implement ObjectMetaAccessor interface."),
+					},
+				}
+			} else {
+				if isNsEvent {
+					pw.k8sNSRev = oma.GetObjectMeta().GetResourceVersion()
+				} else {
+					pw.k8sSARev = oma.GetObjectMeta().GetResourceVersion()
+				}
+				oma.GetObjectMeta().SetResourceVersion(pw.JoinRevisions(pw.k8sNSRev, pw.k8sSARev))
+			}
+		} else if e.Error == nil {
+			log.WithField("event", e).Warning("Event without error or value")
+		}
+
+		// Send the processed event.
+		select {
+		case pw.resultChan <- e:
+			// If this is an error event. check to see if it's a terminating one.
+			// If so, terminate this watcher.
+			if e.Type == api.WatchError {
+				log.WithError(e.Error).Debug("Kubernetes event converted to backend watcher error event")
+				if _, ok := e.Error.(cerrors.ErrorWatchTerminated); ok {
+					log.Info("Watch terminated event")
+					return
+				}
+			}
+
+		case <-pw.context.Done():
+			log.Info("Process watcher done event during watch event in kdd client")
+			return
+		}
+	}
+}
+
+func (c *profileClient) joinRevs(ns, sa string) string {
```

Comment:
Not convinced about these functions.

Firstly, they are specific to the AlphaSA feature, but the name of functions does not indicate that, and these functions are accessible to the whole package not just this file - so a little confusing.

Secondly, I'm not even sure these are being used are they?

I would probably have kept the old SplitNetworkPolicyRevision and JoinNetworkPolicyRevision functions and implemented separately names ones for Profiles.  That way it's a bit easier to follow, and there are fewer changes in code that doesn't need to be impacted by this feature.

Reply:
@robbrockbank good points. 
I will revert to using to distinct SplitNetworkPolicyRevision and JoinNetworkPolicyRevision for NP's and another pair for Profiles that use the alphaSA flag.
For context the ask is to keep the profile revision to be "single version" when the feature flag is off.